### PR TITLE
Add a simple Dask-ML example

### PR DESCRIPTION
**Pull request by AnesBenmerzoug**
This PR adds a simple Dask-ML example and should address issue #459 

It uses the Iris dataset from sklearn and distributes it across workers.

### CODE REVIEW

File: examples/dask_ml_simple.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+# FYI: Objective functions can take additional arguments
+# (https://optuna.readthedocs.io/en/stable/faq.html#objective-func-additional-args).
+def objective(trial):
+    iris = load_iris()
+    X, y = iris.data, iris.target
+    X, y = da.from_array(X, chunks=len(X) // 5), da.from_array(y, chunks=len(y) // 5)
+
+    solver = trial.suggest_categorical('solver', ['admm', 'gradient_descent', 'lbfgs'])
+    C = trial.suggest_uniform('C', 0.0, 1.0)
```
Comment (sile) [inline examples/dask_ml_simple.py]: For showing the "define-by-run" feature of Optuna, how about adding a suggestion of a conditional parameter as follows?
```python
if solver == 'admm' or solver == 'lbfgs':
    penalty = trial.suggest_categorical('penalty', ['l1', 'l2'])
else:
    # 'penalty' parameter isn't relevant for this solver, so we always specify 'l2' as the dummy value.
    penalty = 'l2'

...

classifier = LogisticRegression(max_iter=max_iter, solver=solver, C=C, penalty=penalty)
```

File: setup.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def get_extras_require():
         'checking': ['autopep8', 'hacking'],
         'testing': [
             'pytest', 'mock', 'bokeh', 'plotly', 'chainer>=5.0.0', 'xgboost', 'mpi4py', 'lightgbm',
-            'keras', 'mxnet', 'scikit-optimize', 'tensorflow', 'cma'
+            'keras', 'mxnet', 'scikit-optimize', 'tensorflow', 'cma', 'dask-ml', 'dask[dataframe]'
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply setup.py]: Ah, ok. I didn't notice that.
I will fix it

File: examples/dask_ml_simple.py (hunk 1 above)
Comment (AnesBenmerzoug) [reply examples/dask_ml_simple.py]: I had the exact same code and removed it because when using 'l1' with 'lbfgs' it gives me an error about it being non-differentiable at 0. I will try again with 'admm' and maybe add another solver.

File: examples/dask_ml_simple.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    score = classifier.score(X_test, y_test)
+    return score
+
+
+if __name__ == '__main__':
+    import optuna
+    study = optuna.create_study(direction='maximize')
+
+    # This is used to initialize the workers that will be used by Dask-ML
+    cluster = LocalCluster(n_workers=4, threads_per_worker=2, local_directory='/tmp')
```
[the remark this answers is not in the document]
Comment (AnesBenmerzoug) [reply examples/dask_ml_simple.py]: I will do it. I completely forgot about windows after using ubuntu for so long.

File: examples/dask_ml_simple.py (hunk 1 above)
Comment (sile) [reply examples/dask_ml_simple.py]: Thank you!

File: setup.py [hunk 4]
```diff
@@ -55,7 +55,7 @@ def get_extras_require():
         'checking': ['autopep8', 'hacking'],
         'testing': [
             'pytest', 'mock', 'bokeh', 'plotly', 'chainer>=5.0.0', 'xgboost', 'mpi4py', 'lightgbm',
-            'keras', 'mxnet', 'scikit-optimize', 'tensorflow', 'cma'
+            'keras', 'mxnet', 'scikit-optimize', 'tensorflow', 'cma',
```
Comment (sile) [inline setup.py]: Please remove the trailing comma to eliminate this unnecessary diff.
```suggestion
            'keras', 'mxnet', 'scikit-optimize', 'tensorflow', 'cma'
```

Comment (AnesBenmerzoug) [reply setup.py]: Done

File: examples/dask_ml_simple.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+def objective(trial):
+    iris = load_iris()
+    X, y = iris.data, iris.target
+    X, y = da.from_array(X, chunks=len(X) // 5), da.from_array(y, chunks=len(y) // 5)
+
+    solver = trial.suggest_categorical('solver', ['admm', 'gradient_descent', 'proximal_grad'])
+    C = trial.suggest_uniform('C', 0.0, 1.0)
+
+    if solver == 'admm' or solver == 'proximal_grad':
+        penalty = trial.suggest_categorical('penalty', ['l1', 'l2', 'elastic_net'])
```
Comment (sile) [inline examples/dask_ml_simple.py]: [PERSONAL NOTE] Inside dask-ml, [`Regularizer.get(penalty)`](https://github.com/dask/dask-glm/blob/master/dask_glm/regularizers.py#L125) is called to create an instance of `Regularizer` from the penalty name.

File: examples/dask_ml_simple.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
+"""
+Optuna example that optimizes a classifier configuration for the Iris dataset using Dask-ML.
+
+In this example, we optimize a logistic regression classifier configuration for the Iris dataset.
+The classifier is from dask-ml while the dataset is from sklearn.
+We optimize the choice of solver (admm, gradient descent, or newton),
```
Comment (sile) [inline examples/dask_ml_simple.py]: The objective function in this example seems using `proximal_grad` instead of `newton` as a choice.
```suggestion
We optimize the choice of solver (admm, gradient descent, or proximal gradient),
```

Comment (AnesBenmerzoug) [reply examples/dask_ml_simple.py]: Great catch. I'll fix that

File: examples/dask_ml_simple.py (hunk 5 above)
Comment (AnesBenmerzoug) [reply examples/dask_ml_simple.py]: Do you want me to add that as a comment? I would expect people who want to use Dask-ML to find that in the docs instead

Comment (sile) [reply examples/dask_ml_simple.py]: No, this is just a note for me.

Because [the document of `LogisticRegression`](https://ml.dask.org/modules/generated/dask_ml.linear_model.LogisticRegression.html) doesn't mention to "elastic_net" (it says "For string values, only ‘l1’ or ‘l2’ are valid."), I read the source code to find out the usage of `penalty` string.

Comment (AnesBenmerzoug) [reply examples/dask_ml_simple.py]: I think the docs are not in sync with the code.

[Here](https://github.com/dask/dask-ml/blob/master/dask_ml/linear_model/glm.py#L137,L169) we can see where the penalty attribute is used in LogisticRegression. There is no check for elastic_net and no exception is raised when it is passed.

[That method](https://github.com/dask/dask-ml/blob/master/dask_ml/linear_model/glm.py#L137,L169) is called in the [fit method](https://github.com/dask/dask-ml/blob/master/dask_ml/linear_model/glm.py#L171,L199) and its result is passed to a solver from [this dictionary of solvers](https://github.com/dask/dask-glm/blob/master/dask_glm/algorithms.py#L434).

Since Elastic Net is just the weighted average of L1 and L2 it should work of both L1 and L2 work.

File: examples/dask_ml_simple.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    if solver == 'admm' or solver == 'proximal_grad':
+        penalty = trial.suggest_categorical('penalty', ['l1', 'l2', 'elastic_net'])
+    else:
+        # 'penalty' parameter isn't relevant for this solver,
+        # so we always specify 'l2' as the dummy value.
+        penalty = 'l2'
+
+    max_iter = 100
+
+    classifier = LogisticRegression(max_iter=max_iter, solver=solver, C=C, penalty=penalty)
```
Comment (toshihikoyanase) [inline examples/dask_ml_simple.py]: I think the variable `max_iter` is only used here.
Please input the literal value like `LogisticRegression(max_iter=100, ...`.

File: examples/dask_ml_simple.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    return score
+
+
+if __name__ == '__main__':
+    import optuna
+    study = optuna.create_study(direction='maximize')
+
+    # This is used to initialize the workers that will be used by Dask-ML.
+    cluster = LocalCluster(n_workers=2)
+    with Client(cluster) as client:
```
Comment (toshihikoyanase) [inline examples/dask_ml_simple.py]: Can we remove the variable `client`? Or, is it required by some background procedures of Dask-ML?
I think it is not used in this file.

File: examples/dask_ml_simple.py (hunk 7 above)
Comment (AnesBenmerzoug) [reply examples/dask_ml_simple.py]: I went ahead and made it a parameter to optimize

File: examples/dask_ml_simple.py (hunk 8 above)
Comment (AnesBenmerzoug) [reply examples/dask_ml_simple.py]: It is useful to control the number of worker processes and/or threads.
But it is not necessary in this case because dask-ml will create one internally if there
isn't one already.
I went ahead and removed it

Comment (toshihikoyanase) [reply examples/dask_ml_simple.py]: I understand. Thank you for your response!

File: examples/dask_ml_simple.py (hunk 7 above)
Comment (toshihikoyanase) [reply examples/dask_ml_simple.py]: Thank you for your modification. But, I think that tuning of `max_iter` can be eliminated.
IMO, LogisticRegression stops the training based on the convergence criteria ( `tol` ) and sufficient large `max_iter` usually archives good results with less time compared to tuning iterations parameter.

Comment (AnesBenmerzoug) [reply examples/dask_ml_simple.py]: Ok, I'll change that

Comment (toshihikoyanase) [reply examples/dask_ml_simple.py]: Thanks!
